src/jit/lclvars.cpp

```diff
@@ -3445,84 +3387,86 @@ void Compiler::lvaSortByRefCount()
         if (opts.MinOpts() && compHndBBtabCount > 0)
         {
             lvaSetVarDoNotEnregister(lclNum DEBUGARG(DNER_LiveInOutOfHandler));
-            continue;
         }
-
-        var_types type = genActualType(varDsc->TypeGet());
-
-        switch (type)
+        else
         {
+            var_types type = genActualType(varDsc->TypeGet());
+
+            switch (type)
+            {
 #if CPU_HAS_FP_SUPPORT
-            case TYP_FLOAT:
-            case TYP_DOUBLE:
+                case TYP_FLOAT:
+                case TYP_DOUBLE:
 #endif
-            case TYP_INT:
-            case TYP_LONG:
-            case TYP_REF:
-            case TYP_BYREF:
+                case TYP_INT:
+                case TYP_LONG:
+                case TYP_REF:
+                case TYP_BYREF:
 #ifdef FEATURE_SIMD
-            case TYP_SIMD8:
-            case TYP_SIMD12:
-            case TYP_SIMD16:
-            case TYP_SIMD32:
+                case TYP_SIMD8:
+                case TYP_SIMD12:
+                case TYP_SIMD16:
+                case TYP_SIMD32:
 #endif // FEATURE_SIMD
-            case TYP_STRUCT:
-                break;
+                case TYP_STRUCT:
+                    break;
 
-            case TYP_UNDEF:
-            case TYP_UNKNOWN:
-                noway_assert(!"lvType not set correctly");
-                varDsc->lvType = TYP_INT;
+                case TYP_UNDEF:
+                case TYP_UNKNOWN:
+                    noway_assert(!"lvType not set correctly");
+                    varDsc->lvType = TYP_INT;
 
-                __fallthrough;
+                    __fallthrough;
 
-            default:
-                varDsc->lvTracked = 0;
+                default:
+                    varDsc->lvTracked = 0;
+            }
         }
-    }
-
-    /* Now sort the variable table by ref-count */
-
-    lvaSortOnly();
-
-    /* Decide which variables will be worth tracking */
 
-    if (lvaCount > lclMAX_TRACKED)
-    {
-        /* Mark all variables past the first 'lclMAX_TRACKED' as untracked */
-
-        for (lclNum = lclMAX_TRACKED; lclNum < lvaCount; lclNum++)
+        if (varDsc->lvTracked)
         {
-            lvaRefSorted[lclNum]->lvTracked = 0;
+            tracked[trackedCount++] = lclNum;
         }
     }
 
-    if (lvaTrackedToVarNum == nullptr)
+    // Now sort the tracked variable table by ref-count
+    if (compCodeOpt() == SMALL_CODE)
+    {
+        auto less = [this](unsigned n1, unsigned n2) -> bool { return RefCntCmp(lvaGetDesc(n1), lvaGetDesc(n2)) < 0; };
+        jitstd::sort(tracked, tracked + trackedCount, less);
+    }
+    else
     {
-        lvaTrackedToVarNum = new (getAllocator(CMK_LvaTable)) unsigned[lclMAX_TRACKED];
+        auto less = [this](unsigned n1, unsigned n2) -> bool {
+            return WtdRefCntCmp(lvaGetDesc(n1), lvaGetDesc(n2)) < 0;
+        };
+        jitstd::sort(tracked, tracked + trackedCount, less);
```

Comment:
I do not understand how do you decide which variables are worth tracking without sorting all of them?

Reply:
Well, I do sort all variables that can be tracked, just not all (trackable + untrackable) variables. And below the `tracked` array is effectively "truncated" to keep at most `lclMAX_TRACKED` variables.

It would be even better to use selection instead of sorting but we don't seem to hit the tracking limit often and when we do the numbers aren't high enough for `O(n log k)` to be much different from `O(n log n)`.

Reply:
Got it, I missed that part. thanks for the explanation.